code/dynamic_programming/subset_sum/subset_sum.cpp

```diff
@@ -1,44 +1,76 @@
 /* Part of Cosmos by OpenGenus Foundation */
+#ifndef SUBSET_SUM
+#define SUBSET_SUM
 
-#include <iostream>
-#include <cstring>
+#include <iterator>
 
-#define MAX 110
-#define MAX_SUM 100100
+/*
+ *  Check whether is possible to
+ *  get value sum from a subset
+ *  of the [begin:end)
+ */
 
-int reachable[MAX_SUM];
-bool dp[MAX_SUM];
+// complexity: time: O(sum * n), space: O(sum)
+template<typename _BidirectionalIter,
+         typename _ValueType = typename std::iterator_traits<_BidirectionalIter>::value_type>
+bool
+isSubsetSum(_BidirectionalIter begin, _BidirectionalIter end, _ValueType sum)
+{
+    auto sz = std::distance(begin, end);
+    bool subset[2][sum + 1];
```

Comment:
Any reason why you are using a 2D raw array?

Reply:
[Refer to the description of this article.](http://www.geeksforgeeks.org/subset-sum-problem-osum-space/)